src/test/arith_uint256_tests.cpp

```diff
@@ -571,4 +552,50 @@ BOOST_AUTO_TEST_CASE( getmaxcoverage ) // some more tests just to get 100% cover
     CHECKBITWISEOPERATOR(R1,~R2,&)
 }
 
+BOOST_AUTO_TEST_CASE(conversion)
```

Comment:
so these are basically roundtrip tests now, right?
These are very valuable (could even be a fuzz test now), but could we add back some of the hardcoded ones to make sure they're not just self-referentially correct (i.e. circular reasoning), i.e. something like:
```C++
    BOOST_CHECK_EQUAL(UintToArith256(ZeroL), 0);
    BOOST_CHECK_EQUAL(UintToArith256(OneL), 1);
    BOOST_CHECK_EQUAL(ArithToUint256(0), ZeroL);
    BOOST_CHECK_EQUAL(ArithToUint256(1), OneL);

    BOOST_CHECK_EQUAL(ArithToUint256(ZeroL), uint256::ZERO);
    BOOST_CHECK_EQUAL(ArithToUint256(OneL), uint256::ONE);
    BOOST_CHECK_EQUAL(UintToArith256(uint256::ZERO), ZeroL);
    BOOST_CHECK_EQUAL(UintToArith256(uint256::ONE), OneL);
```

Reply:
> so these are basically roundtrip tests now, right?

I think `BOOST_CHECK_EQUAL(UintToArith256(ArithToUint256(arith)), arith);` is the only roundtrip?

> but could we add back some of the hardcoded ones to make sure they're not just self-referentially correct (i.e. circular reasoning):

Isn't that covered here? https://github.com/bitcoin/bitcoin/pull/30773/files#diff-d06d0fcd312512f190cc0d05450dd55973c60aeb8d57ed2d705e7beaa383f386R565-R568

Reply:
Not exactly, this looks more like a trampoline than simple hard-coded examples for just `ArithToUint256` and `UintToArith256`. Both sides depend on the same parameter, which seems circular to me-at least from a black-box perspective. However, I’m fine with this if you think it’s sufficient.

Reply:
I'm sorry, I just don't see how those hardcoded tests are any different from the ones I linked. `ZeroL` is the same as `arith_uint256{0}`, and `uint256::ZERO` is the same as `uint256{0}`. Besides these aliases being irrelevant for a conversion test, their equivalence is tested in other unit tests already. What am I missing?

Reply:
You know at the implementation, that's why it seems the same.
The test would also pass for a simple re-cast roundtrip:
```C++
auto ArithToUint256_mock(const arith_uint256 &a) { return *reinterpret_cast<const uint256*>(&a); }
auto UintToArith256_mock(const uint256 &a) { return *reinterpret_cast<const arith_uint256*>(&a); }
BOOST_AUTO_TEST_CASE(conversion)
{
    for (const arith_uint256& arith : {ZeroL, OneL, R1L, R2L}) {
        const auto u256{uint256::FromHex(arith.GetHex()).value()};
        BOOST_CHECK_EQUAL(UintToArith256_mock(ArithToUint256_mock(arith)), arith);
        BOOST_CHECK_EQUAL(UintToArith256_mock(u256), arith);
        BOOST_CHECK_EQUAL(u256, ArithToUint256_mock(arith));
        BOOST_CHECK_EQUAL(ArithToUint256_mock(arith).GetHex(), UintToArith256_mock(u256).GetHex());
    }

    for (uint8_t num : {0, 1, 0xff}) {
        BOOST_CHECK_EQUAL(UintToArith256_mock(uint256{num}), arith_uint256{num});
        BOOST_CHECK_EQUAL(uint256{num}, ArithToUint256_mock(arith_uint256{num}));
    }
}
```

That's why I though we should cover it with hard-coded tests which cannot cheat by passing the parameter around.
(note, the previous hard-coded ones also pass for the above `_mock` methods, we should find better ones to obviate why those are incorrect)

Reply:
> You already know at the implementation

At which points are these test relying on implementation?

Thanks for providing an example. If you have a diff that shows where the "hardcoded" tests catch something that "my" tests don't, that would really help. I struggle to see how `BOOST_CHECK_EQUAL(ArithToUint256(ZeroL), uint256::ZERO);` is any different from `BOOST_CHECK_EQUAL(ArithToUint256(arith_uint256{0}), uint256{0});` (or how that substitution is relying on implementation details, which you seem to be hinting at).

Reply:
> The test would also pass for a simple re-cast roundtrip:

I'm a bit confused by the example `_mock` functions. I think reinterpret casting a `std::array<uint8_t, 32>` <-> `uint32_t[8]` can actually be a (unreliable and unsafe but depending on memory alignment sometimes correct) way of implementing the conversion functions, since both types are 256 contiguous bits of uint data? So depending on your platform, I think this passing the tests doesn't really provide much information?

Sorry if I'm being dim, I'm really trying to understand your concern.

Reply:
> Sorry if I'm being dim, I'm really trying to understand your concern.

I can see that and I appreciate it.

> I think BOOST_CHECK_EQUAL(UintToArith256(ArithToUint256(arith)), arith); is the only roundtrip?

Aren't these roundtrip as well because of `GetHex`/`FromHex`?
```C++
const auto u256{uint256::FromHex(arith.GetHex()).value()};
BOOST_CHECK_EQUAL(UintToArith256(u256), arith);
BOOST_CHECK_EQUAL(u256, ArithToUint256(arith));
BOOST_CHECK_EQUAL(ArithToUint256(arith).GetHex(), UintToArith256(u256).GetHex());
```

> I think this passing the tests doesn't really provide much information?

It's not trivial to find a counter example here which would still pass - but that's exactly my concern, that's it's too complicated, that if e.g. `GetHex` would accidentally zero out the internal state, many of these would pass, but comparing against a constant value is harder to trick, since only one side can change (easier to grasp than nested methods and parameters and states).

E.g. imagine having a trim method, and we're asserting that when it doesn't start with a space, the length is the same as the original (`BOOST_CHECK_EQUAL(Trim(orig).size(), orig.size());`). Would you trust this `Trim` method without having a concrete example like `BOOST_CHECK_EQUAL(Trim("abc"), "abc");` and `BOOST_CHECK_EQUAL(Trim(" abc "), "abc");`, etc?

In fuzz testing the above ones are fine, since we're checking invariants (i.e. constant properties) of the method (e.g. that a trimmed method's size can never be greater than the original), but in unit testing we usually provide simple examples that serve as documentation as well.

So all in all, I think the above examples are very good property-based tests, but I'm missing simple unit tests that can document the behavior of the method in a trivial way (i.e. when they fail I want to pinpoint the reason quickly).

But I've already ACK-ed this, it's not a blocker, if this is still doesn't make sense, please mark it as resolved, it's not that important :)

Reply:
> but I'm missing simple unit tests that can document the behavior of the method in a trivial way

If we assume that `GetHex()` is a complete representation of each class' internal state, then feel like it's reasonable (even if suboptimal) to rely on using that method do assess the correctness of the conversion function, if `GetHex()` is well unit tested. Also note that we have tests that don't rely on `GetHex()` / `FromHex()` at all, i.e.:

```
    for (uint8_t num : {0, 1, 0xff}) {
        BOOST_CHECK_EQUAL(UintToArith256(uint256{num}), arith_uint256{num});
        BOOST_CHECK_EQUAL(uint256{num}, ArithToUint256(arith_uint256{num}));
        BOOST_CHECK_EQUAL(UintToArith256(uint256{num}), num);
    }
```

That said, I think we both agree that having trivial checks that rely as little as possible on other functions is good. I think the `for (uint8_t num : {0, 1, 0xff}) {` tests are a good example of that, but they could be expanded by constructing more `uint256`'s from hardcoded arrays. I think my confusion arose from the fact that we already _are_ doing such "hardcoded" conversion tests (for `{0, 1, 0xff}`), but this could be improved upon by adding more `uint256`s that are larger than 0xff?

While I don't object to that, I think it's not really necessary to this PR, so since it's not blocking for you I think it makes sense to leave it at this.